### PR TITLE
Fused BatchNormAddRelu

### PR DESCRIPTION
**Pull request by liujuncheng**
添加`flow.layers.batch_normalization_add_relu`和`flow.layers.batch_normalization_relu`接口，在符合条件(NHWC，FP16，channel维是4的倍数)时kernel将使用cuDNN的融合api以提高性能，其他条件下性能和原来基本相近

### CODE REVIEW

File: oneflow/user/ops/normalization_op.cpp [hunk 1]
```diff
@@ -342,4 +542,68 @@ REGISTER_USER_OP_GRAD("normalization")
                                 });
     });
 
+REGISTER_USER_OP_GRAD("normalization_add_relu")
```
Comment (leaves-zwx) [inline oneflow/user/ops/normalization_op.cpp:545]: 不需要 "cudnn_fused_normalization_add_relu" 的 grad op 注册吗？

File: oneflow/user/ops/normalization_op.cpp [hunk 2]
```diff
@@ -150,23 +322,51 @@ REGISTER_USER_OP("normalization_grad")
...
+    .Input("mean")
+    .Input("inv_variance")
+    .Input("gamma")
+    .Input("beta")
+    .Input("reserve_space")
+    .Input("y")
+    .Output("gamma_diff")
+    .Output("beta_diff")
+    .Output("dx")
+    .OptionalOutput("d_addend")
```
Comment (leaves-zwx) [inline oneflow/user/ops/normalization_op.cpp:341]: 这个按照惯例是不是叫 addend_diff ？

File: oneflow/core/job_rewriter/cudnn_fused_normalization_add_relu_pass.cpp [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    if (op_type_name != "normalization_add_relu" && op_type_name != "normalization_add_relu_grad") {
+      return;
+    }
+    const user_op::UserOpConfWrapper user_op_conf(op_conf);
+    const BlobDesc& x_desc =
+        op_node->LogicalBlobDesc4Lbi(GenLogicalBlobId(user_op_conf.input("x", 0)));
+    const int32_t axis = user_op_conf.attr<int32_t>("axis");
+    if (x_desc.data_type() != DataType::kFloat16) { return; }
+    const Shape& x_shape = x_desc.shape();
+    if (x_shape.Count(axis + 1) != 1) { return; }
```
Comment (leaves-zwx) [inline oneflow/core/job_rewriter/cudnn_fused_normalization_add_relu_pass.cpp:66]: 这个判断是保证 format 为 NHWC 吗？

Comment (liujuncheng) [reply oneflow/core/job_rewriter/cudnn_fused_normalization_add_relu_pass.cpp:66]: 是的

File: oneflow/user/ops/normalization_op.cpp (hunk 1 above)
Comment (liujuncheng) [reply oneflow/user/ops/normalization_op.cpp:547]: 不需要，用户插入的和前后向展开的是normalization_add_relu和normalization_add_relu_grad op，然后会有一个pass判断符合条件的替换为cudnn_fused_normalization_add_relu和cudnn_fused_normalization_add_relu_grad

File: oneflow/user/ops/normalization_op.cpp (hunk 2 above)
Comment (liujuncheng) [reply oneflow/user/ops/normalization_op.cpp]: 好的

File: oneflow/user/ops/normalization_op.cpp (hunk 1 above)
Comment (leaves-zwx) [reply oneflow/user/ops/normalization_op.cpp:547]: 嗯嗯，CudnnFusedNormalizationAddReluPass 这个 pass 既替换了前向又替换了后向。

File: oneflow/python/ops/layers.py [hunk 4]
```diff
@@ -1182,6 +1213,188 @@ def update_moving(moving, this_batch):
...
+        .Attr("epsilon", epsilon)
+        .Attr("momentum", momentum)
+    )
+    if addend is not None:
+        builder = builder.Input("addend", [addend])
+    return builder.Build().InferAndTryRun().RemoteBlobList()[0]
+
+
+@oneflow_export("layers.batch_normalization_relu")
+def batch_normalization_relu(
```
Comment (daquexian) [inline oneflow/python/ops/layers.py:1336]: 用户需要自己写 `flow.layers.batch_normalization_add_relu` 或 `flow.layers.batch_normalization_relu` 吗，为什么不是自动做 op fusion 呢 0。0

Comment (liujuncheng) [reply oneflow/python/ops/layers.py:1336]: > 用户需要自己写 `flow.layers.batch_normalization_add_relu` 或 `flow.layers.batch_normalization_relu` 吗，为什么不是自动做 op fusion 呢 0。0

后续可以支持

File: oneflow/user/kernels/normalization_kernel.cu [hunk 5]
```diff
@@ -0,0 +1,726 @@
...
+  CudnnTensorDescHelper(const ShapeView& xy_shape, const DataType& data_type, const int32_t axis,
+                        cudnnBatchNormMode_t mode) {
+    OF_CUDNN_CHECK(cudnnCreateTensorDescriptor(&xy_desc_));
+    InferXYCudnnTensorDesc(xy_shape, data_type, axis, xy_desc_);
+    OF_CUDNN_CHECK(cudnnCreateTensorDescriptor(&param_desc_));
+    InferParamCudnnTensorDesc(xy_desc_, mode, param_desc_);
+    int n, c, h, w, n_stride, c_stride, h_stride, w_stride;
+    OF_CUDNN_CHECK(cudnnGetTensor4dDescriptor(param_desc_, &param_data_type_, &n, &c, &h, &w,
+                                              &n_stride, &c_stride, &h_stride, &w_stride));
+    param_size_ = c;
```
Comment (leaves-zwx) [inline oneflow/user/kernels/normalization_kernel.cu:78]: 这里是假定了 `cudnnBatchNormMode_t` 不会为 `CUDNN_BATCHNORM_PER_ACTIVATION`, 所以 param 的形状一定为 1xCx1x1 吗？

File: oneflow/user/kernels/normalization_kernel.cu [hunk 6]
```diff
@@ -0,0 +1,726 @@
...
+
+REGISTER_BN_TRAIN_KERNEL(float16)
+REGISTER_BN_TRAIN_KERNEL(float)
+REGISTER_BN_TRAIN_KERNEL(double)
+
+#define REGISTER_BN_ADD_RELU_KERNEL(dtype)                                            \
+  REGISTER_USER_KERNEL("normalization_add_relu")                                      \
+      .SetCreateFn<NormalizationTrainKernel<dtype>>()                                 \
+      .SetIsMatchedHob((user_op::HobDeviceTag() == "gpu")                             \
+                       & (user_op::HobDataType("y", 0) == GetDataType<dtype>::value)) \
```
Comment (leaves-zwx) [inline oneflow/user/kernels/normalization_kernel.cu:351]: 这里为啥没有 `(user_op::HobAttr<bool>("training") == true))`

Comment (leaves-zwx) [reply oneflow/user/kernels/normalization_kernel.cu:351]: 明白了，`normalization_add_relu` 没有 `training` attr，它默认就是 training

File: oneflow/user/kernels/normalization_kernel.cu (hunk 5 above)
Comment (liujuncheng) [reply oneflow/user/kernels/normalization_kernel.cu:78]: > 这里是假定了 `cudnnBatchNormMode_t` 不会为 `CUDNN_BATCHNORM_PER_ACTIVATION`, 所以 param 的形状一定为 1xCx1x1 吗？

是的

File: oneflow/user/ops/normalization_op.cpp [hunk 7]
```diff
@@ -150,23 +324,51 @@ REGISTER_USER_OP("normalization_grad")
...
+    .SetBatchAxisInferFn(BwBatchAxisInferFn)
+    .SetGetSbpFn(BwGetSbpFn);
+
+REGISTER_USER_OP("normalization_add_relu_grad")
+    .Input("x")
+    .Input("dy")
+    .Input("mean")
+    .Input("inv_variance")
+    .Input("gamma")
+    .Input("beta")
```
[the remark this answers is not in the document]
Comment (liujuncheng) [reply oneflow/user/ops/normalization_op.cpp:337]: > `Input("beta")` 是不是只有在 cudnn_fused 版本的 op 里面才用的着，在这个 `normalization_add_relu_grad` 里没有用，只是为了保持接口统一才设置的？

是的

**Review by leaves-zwx — APPROVED** [review]

**Review by guo-ran — APPROVED** [review]
